### ISSUE TITLE
Misplaced attributes for deprecation markers

### ISSUE THREAD

**Issue by jonludlam** [ocaml/ocaml#13054]
I've just been testing odoc's support for 5.2 as contributed by @Octachron, which is working well and passing tests. However, one of our tests is producing a new warning that it wasn't before. The code in question looks like this:

```ocaml
...
module Top1 : sig
  [@@@deprecated "A"]

  (** Top-comment. *)
end
...
```

and I'm now getting a misplaced attribute warning:

```
File "cases/alerts.mli", line 20, characters 6-16:
20 |   [@@@deprecated "A"]
           ^^^^^^^^^^
Warning 53 [misplaced-attribute]: the "deprecated" attribute cannot appear in this context
```

I saw that @ccasin did a bunch of fixes in this area in #12451 so I had a look through that, but I'm afraid it's still not clear to me that this is expected behaviour. In odoc this is only generating a warning at compile time of the test, and the generated HTML still has the deprecated alert in it, so I _think_ we might be missing a whitelisting of this attribute somewhere. If not, what's the right way to mark a whole module as deprecated?

**Comment by Octachron** [comment]
The correct syntax is
```ocaml
module Top1: sig
  ...
end[@@deprecated "A"]
```
or
```ocaml
module[@deprecated "A"] Top1: sig
  ...
end
```
The improved warning is reliable in the sense that the warning is now always produced whenever the compiler ignored a builtin attribute which was not the case before.

**Comment by ccasin** [comment]
> The improved warning is reliable in the sense that the warning is now always produced whenever the compiler ignored a builtin attribute which was not the case before.

This is true for most attributes by design, but for `deprecated` the correctness of w53 is less certain (even after #12451).  For most attributes we can decide not to issue w53 at the moment the compiler "uses" the attribute, but for `deprecated` the compiler "uses" it by just putting it into the environment where it may be found later by a check for issuing a deprecated warning.  I did a fair amount of manual testing to ensure the places where we do/don't issue w53 for `deprecated` line up with the places that aren't/are checked later for deprecated warnings, but it is possible there is a bug here I will be happy to fix if needed.  Such a bug would manifest by, either:

* The compiler issuing w53 for a `deprecated` attribute but then also issuing a `deprecated` warning for the place where this attribute appears, or
* The compiler issuing not issuing w53 for a `deprecated` attribute but also not issuing a `deprecated` warning when you use the definition the attribute is on.

In @jonludlam's original example I believe we are not in either of these cases - the compiler does issue w53 and does not issue a deprecated warning for uses of elements of this signature - but I will not be terribly surprised if I got this wrong.

**Comment by Octachron** [comment]
The example in question is indeed a case where the attribute was ignored.

**Comment by jonludlam** [comment]
> The correct syntax is...

That makes sense for submodules, and we can change our test cases to look like that. I'm curious though, how would you go about deprecating an entire compilation unit?

**Comment by Octachron** [comment]
This is the case where a floating attribute in the "header" of the file (before non-attribute items) works:

```ocaml
[@@@deprecated "ok"]
val x: int
```
```ocaml
val x:int
[@@@deprecated "not_in_the_header"]
```
And there seems to be a warning 53 bug in this case, since the first form trigger a warning `53` in a ml only file whereas the warning doesn't trigger in the second case.

**Comment by gasche** [comment]
(See the related RFC https://github.com/ocaml/RFCs/pull/26 that proposes giving a formal status to this idea of "header of a compilation unit".)

**Comment by ccasin** [comment]
> And there seems to be a warning 53 bug in this case, since the first form trigger a warning `53` in a ml only file whereas the warning doesn't trigger in the second case.

Just wanted to acknowledge I've seen this and plan to fix it.  Sorry for the delay, it's been a busy week.

**Comment by emillon** [comment]
Hi.
To expand on this, this affects dune as well. To support `(wrapped (transition))`, we compile to something that looks like:

```ocaml
(* l.ml *)
[@@@alert example "This is an example"]

let x = 1
(* main.ml *)
let () = Printf.printf "%d\n" L.x
```

When built with `ocamlc l.ml main.ml`, we get:

```
File "l.ml", line 1, characters 4-14:
1 | [@@@deprecated "This is an example"]
        ^^^^^^^^^^
Warning 53 [misplaced-attribute]: the "deprecated" attribute cannot appear in this context

File "main.ml", line 1, characters 30-33:
1 | let () = Printf.printf "%d\n" L.x
                                  ^^^
Alert deprecated: module L
This is an example

File "l.ml", line 1, characters 4-14:
Warning 53 [misplaced-attribute]: the "deprecated" attribute cannot appear in this context
```

So the attribute isn't ignored (because it triggers the alert), but warning 53 still fires.

**Comment by nojb** [comment]
Sorry to come late to the discussion, but I believe this issue was fixed in #13170. Maybe the fix should be back-ported to 5.2? (Also, if someone can confirm that it is indeed fixed, that would also be useful.)

**Comment by emillon** [comment]
Hey @nojb. Thanks. I checked with trunk today and this small reproduction case is indeed fixed, but the test for `(wrapped (transition))` in dune still fails with it. I'll try to come up with a better reproduction case.

**Comment by nojb** [comment]
> Hey @nojb. Thanks. I checked with trunk today and this small reproduction case is indeed fixed, but the test for `(wrapped (transition))` in dune still fails with it. I'll try to come up with a better reproduction case.

I believe the warning is correctly triggered. Indeed, the command line used by Dune to compile the generated "unwrapped" module looks like this:
```
ocamlc.opt -w @1..3@5..28@30..39@43@46..47@49..57@61..62-40 -strict-sequence -strict-formats -short-paths -keep-locs -g -bin-annot -I .foo.objs/byte -no-alias-deps -opaque -o .foo.objs/byte/bar.cmo -c -impl .wrapped_compat/Bar.ml-gen
```
Here `.foo.objs/byte` contains `bar.cmi`, which is used by the compiler as the interface of `Bar.ml-gen`. Under those circumstances, top-level attributes in the `.ml-gen` are indeed ignored by the compiler.

**Comment by nojb** [comment]
> I believe the warning is correctly triggered.

To clarify: Dune first builds `bar.cmi` and `bar.cmo` from `bar.ml-gen` and this part works well (`bar.ml-gen` contains a top-level attribute but no `.cmi`). However, when it is the turn of the native-code compiler, `bar.cmi` already exists (and contains the expected deprecation attribute), so the top-level attribute in `Bar.ml-gen` is correctly ignored by the compiler, hence the warning.

**Comment by emillon** [comment]
Oh, fantastic, you saved me a lot of debugging time. I'll try to fix that on the dune side.

**Comment by nojb** [comment]
I am tentatively closing this issue as it appears it is indeed fixed by #13170. We'll reopen if needed.
